File: lib-segment/src/main/java/websiteschema/mpsegment/graph/SparseVector.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package websiteschema.mpsegment.graph;

import java.util.Iterator;
import java.util.TreeMap;

/**
 *
 * @author twer
 */
public class SparseVector<Obj> {

    private final int N;             // length
    private SymbolTable<Integer, Integer, Obj> st;  // the vector, represented by index-value pairs

    // initialize the all 0s vector of length N
    public SparseVector(int N) {
        this.N = N;
        this.st = new SymbolTable<Integer, Integer, Obj>();
    }

    // put st[i] = value
    public void put(int i, int value, Obj obj) {
        if (i < 0 || i >= N) {
            throw new RuntimeException("Illegal index");
        }
        if (value == 0.0) {
            st.delete(i);
        } else {
            st.put(i, value, obj);
        }
    }

    // return st[i]
    public int get(int i) {
        if (i < 0 || i >= N) {
            throw new RuntimeException("Illegal index");
        }
        if (st.contains(i)) {
            return st.get(i);
        } else {
            return 0;
        }
    }

    public Obj getObject(int i) {
        if (i < 0 || i >= N) {
            throw new RuntimeException("Illegal index");
        }
        if (st.contains(i)) {
            return st.getObject(i);
        } else {
            return null;
        }
    }

    // return the number of nonzero entries
    public int nnz() {
        return st.size();
    }

    public int[] getNonZeroColumns() {
        int cols[] = new int[nnz()];
        Iterator<Integer> keys = st.iterator();
        if (null != keys) {
            int i = 0;
            while (keys.hasNext()) {
                cols[i++] = keys.next();
            }
        }
        return cols;
    }

    // return the size of the vector
    public int size() {
        return N;
    }

    // return a string representation
    @Override
    public String toString() {
        String s = "";
        for (int i : st) {
            s += "(" + i + ", " + st.get(i) + ") ";
        }
        return s;
    }

    void clear() {
        st.clear();
    }

    /**
     * ***********************************************************************
     * Compilation: javac ST.java Execution: java ST
     *
     * Sorted symbol table implementation using a java.util.TreeMap. Does not
     * allow duplicate keys.
     *
     * % java ST
     *
     ************************************************************************
     *
     */
    class SymbolTable<Key extends Comparable<Key>, Value, Obj> implements Iterable<Key> {

        private void clear() {
            st.clear();
        }

        class Pair {

            Value value;
            Obj obj;

            Pair(Value v, Obj o) {
                value = v;
                obj = o;
            }
        }
        private TreeMap<Key, Pair> st;

        /**
         * Create an empty symbol table.
         */
        SymbolTable() {
            st = new TreeMap<Key, Pair>(); // 3 is an optimize number only for mpsegment
        }

        /**
         * Put key-value pair into the symbol table. Remove key from table if
         * value is null.
         */
        void put(Key key, Value val, Obj obj) {
            if (val == null) {
                st.remove(key);
            } else {
                st.put(key, new Pair(val, obj));
            }
        }

        /**
         * Return the value paired with given key; null if key is not in table.
         */
        Value get(Key key) {
            return st.get(key).value;
        }

        Obj getObject(Key key) {
            return st.get(key).obj;
        }

        Value delete(Key key) {
            return st.remove(key).value;
        }

        boolean contains(Key key) {
            return st.containsKey(key);
        }

        int size() {
            return st.size();
        }

        @Override
        public Iterator<Key> iterator() {
            return st.keySet().iterator();
        }
    }
}

File: lib-segment/src/main/scala/me/yingrui/segment/graph/SparseVector.scala
package me.yingrui.segment.graph

import java.util

class SparseVector[Obj](N: Int) {

  // initialize the all 0s vector of length N
  private val st: SymbolTable[Int, Obj] = new SymbolTable[Int, Obj]() // the vector, represented by index-value pairs

  // put st(i) = value
  def put(i: Int, value: Int, obj: Obj) {
    if (i < 0 || i >= N) {
      throw new RuntimeException("Illegal index")
    }
    if (value == 0) {
      st.delete(i)
    } else {
      st.put(i, value, obj)
    }
  }

  // return st(i)
  def get(i: Int): Int = {
    if (i < 0 || i >= N) {
      throw new RuntimeException("Illegal index")
    }
    return st.getOrElse(i, 0)
  }

  def getObject(i: Int): Option[Obj] = {
    if (i < 0 || i >= N) {
      throw new RuntimeException("Illegal index")
    }

    st.getOrElseObject(i)
  }

  // return the number of nonzero entries
  def nnz(): Int = {
    return st.size
  }

  def getNonZeroColumns(): Array[Int] = {
    val cols = new Array[Int](nnz())
    val keys = st.iterator()
    if (null != keys) {
      var i = cols.length - 1
      while(i >= 0){cols(i) = keys.get(i); i -= 1}
    }
    return cols
  }

  // return the size of the vector
  def size(): Int = {
    return N
  }

  // return a string representation
  override def toString(): String = {
    var s = ""
    val keys = st.iterator()
    var i = 0
    while (i < keys.size()) {
      s += "(" + i + ", " + st.get(i) + ") "
      i += 1
    }
    return s
  }

  def clear() {
    st.clear()
  }

  /**
   * ***********************************************************************
   *
   * Sorted symbol table implementation using a HashMap. Does not
   * allow duplicate keys.
   *
   * ***********************************************************************
   *
   */
  class SymbolTable[Value, Obj] {

    def clear() {
      st.clear()
    }

    class Pair(value: Value, obj: Obj) {
      def getValue() = value

      def getObj() = obj
    }

    /**
     * Create an empty symbol table.
     */
    private val st = new java.util.TreeMap[Int, Pair]()


    /**
     * Put key-value pair into the symbol table. Remove key from table if
     * value is null.
     */
    def put(key: Int, valu: Value, obj: Obj) {
      if (valu == null) {
        st.remove(key)
      } else {
        st.put(key, new Pair(valu, obj))
      }
    }

    /**
     * Return the value paired with given key; null if key is not in table.
     */
    def get(key: Int): Value = {
      return st.get(key).getValue()
    }

    def getOrElse(key: Int, defaultValue: Value): Value = {
      val pair = st.get(key)
      if (null != pair) pair.getValue() else defaultValue
    }

    def getObject(key: Int): Obj = {
      return st.get(key).getObj()
    }

    def getOrElseObject(key: Int): Option[Obj] = {
      val pair = st.get(key)
      if (pair != null) {
        return Option(pair.getObj())
      }
      return None
    }

    def delete(key: Int) {
      st.remove(key)
    }

    def contains(key: Int): Boolean = {
      return st.containsKey(key)
    }

    def size() = st.size

    def iterator(): java.util.List[Int] = {
      new java.util.ArrayList[Int](st.keySet())
    }
  }

}
